### DISCUSSION TITLE
Not maintaining speed

### DISCUSSION THREAD

**Discussion by hudkinsnoah** [JSBSim-Team/jsbsim#1056] [Q&A]
Hello! I feel like there is something that I am missing as I am working here but not sure exactly where to look. I currently have an implementation of a flightsim using the c310 model which fly's just fine with my joystick, autopilot controls, etc. I tried leaving everything the exact same except for the model that I am using. I have tried a couple, but the main one that I have played around with is the DHC6. I changed out the model loaded and tried flying. I am now seeing that the plane is losing speed and starts to fall out of the sky. It feels like the engines aren't on but I can't find a property to check or set. Things work just fine with the c310, but now running into this issue. 

I can see that the "propulsion/starter_cmd" property is set to 1 and I have tried ensuring that a good portion of the references I am using are valid or at the very least in the property catalog and it appears that they are. Any suggestions on what I might be missing would be greatly appreciated!

**Comment by hudkinsnoah** [comment]
Update to this, for some reason the aircraft is holding speed, but it seems to be slow enough that it can't keep itself in the air. Averaging about 108-110 knots which doesn't seem like it is enough to stay in the air and slowly loses elevation

**Reply by hudkinsnoah** [reply]
I have noticed this also with the T37 model. With that model, there is no minimum elevation so I can't ever hit the ground, I just continue falling and can have elevations from the property "atmosphere/pressure-altitude" of -23000 ft

**Comment by seanmcleod** [comment]
Who is controlling the thrust? An auto-pilot or are you as the pilot controlling thrust with a thrust lever input?

Plot things like airspeed, altitude, pitch angle, flight path angle, AoA, thrust, thrust command etc.

**Reply by hudkinsnoah** [reply]
I am controlling the thrust as the pilot with thrust input. I will plot those things and get back to you with that information!

**Reply by hudkinsnoah** [reply]
When I am using the T37 model, I see the unexpected behavior of two things. One, the elevation can dip well below 0 and can fly at negative altitudes (shouldn't be able to go below 0 at the very least). The second, is despite being given a thrust, the speed continues to decrease (unless I am pitched down towards the ground). This is the plot of my altitude using the property "position/h-agl-ft": 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/615fa471-65a4-4213-a119-51d9fea6cc52)
We can see that we end up going below 0 and it never actually bounces or goes back up, just continue heading down.

This is the plot of my pitch using the property "attitude/pitch-rad". Values are in radians: 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/3c9bebf4-b8f6-4d4c-8a5d-b840d19e856c)
I was trying to manually keep the plane at about level or pitched slightly down to see if the speed would increase. It did not and the elevation continued to fall. 

This is the plot of the thrust command which remained at 1 except for a moment where I dropped it quickly from 1 to 0 and back as I have seen my joystick occasionally feed 0 despite being at 1 because it hasn't moved: 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/4b478ffe-1b27-4a95-9526-0e6bfe33ec36)

This is the plot of the airspeed which is found from the property "velocities/vtrue-kts": 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/0a3f7c1c-d32a-4f37-af34-61bd7725a90f) 
The speed only starts to increase because I have pitched down and the aircraft picks up speed as it heads down.


The following data is what I get with the C310 which has the expected behavior: The altitude using the same property (position/h-agl-ft) increases or maintains as I keep the pitch fairly even. When we hit 0, we bounce and head back up again, even with the pitch being down.
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/16da4ab3-194c-43d2-98ff-2a04c528a51c)

The pitch remains fairly steady for a while and then drops to have us sent towards the ground.
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/9430fc0d-b23a-437b-bc5c-8ec43c59214e)

The thrust is the same as before.
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/b47bc67d-f4eb-431b-85c8-83f4d3b85a58)

The airspeed is increasing whether we are remaining steady or even slightly above 0 for pitch. You can also see that the speed is reactive to the throttle being taken down to 0 and put back to 1 again:
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/b55471bb-c0fd-471a-993c-9a34418606a0)
It does initially go down as 200 is a little high for that model, but it recovers and starts to slowly increase or hold steady. 

I have noticed this same pattern on the DCH6 aircraft as I have with the T37. Hopefully this helps with understanding the issue!

**Comment by seanmcleod** [comment]
General comment, in aviation elevation refers to the height of the terrain. 

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/c8afd9d0-44cc-4f1e-b5e1-7dd3150f9829)

In terms of the aircraft flying below sea-level if you look at the T-37 you'll notice that the only ground reactions that are defined are for the landing gear which is retractable. So if the gear is retracted then there will be no impact with the terrain.

https://github.com/JSBSim-Team/jsbsim/blob/b9bfc2350e0e00e53246b65979cc18df2922a88d/aircraft/T37/T37.xml#L91-L137

You'll notice that other aircraft, especially if they have retractable gear will also define structural contact points, for example:

https://github.com/JSBSim-Team/jsbsim/blob/b9bfc2350e0e00e53246b65979cc18df2922a88d/aircraft/787-8/787-8.xml#L201-L214

So in this case even if the gear is retracted one of the structural points will impact with the terrain.

You've shown the position of the thrust lever, but that doesn't mean anything if the engine isn't running. Hence my comment to also plot the actual thrust force to confirm that the engine(s) are actually running.

**Reply by hudkinsnoah** [reply]
Thanks for the diagram and I apologize for my misuse of some of terminology. I am not a pilot and so sometimes I use those interchangeably when they shouldn't be 😃 I will do my best to clarify those better in the future. 

I plotted the thrust pounds (I think that is the correct property: propulsion/engine/thrust-lbs) and the c310 which was working properly has data: 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/d3efe15a-28e5-4165-8a84-02f7e25104b9)

The T37 is consistent at 0 except for an initial high value: 
![image](https://github.com/JSBSim-Team/jsbsim/assets/81835412/77de3bf6-26d7-4385-a733-17b1ea497d90)

Is there something specific that I need to look at for starting the engines?

Putting the landing gear down did result in an impact with the ground rather than continuing to fly at negative altitudes. I will add in those ground reactions in order to make the aircraft impact without dependency on the landing gear.

**Comment by seanmcleod** [comment]
> I apologize for my misuse of some of terminology

No need to apologize.

> Is there something specific that I need to look at for starting the engines?

Couple of options, depending on how you're using JSBSim.

So first off you can specify that the engines are running in the initialisation file, the -1 means all engines.

```xml
<?xml version="1.0" encoding="utf-8"?>
<initialize name="cruise">
  <!--
    This file sets up the aircraft to start off
    at cruise conditions for the rudder kick.
  -->
  <altitude unit="FT">  1000.0 </altitude>
  <vc unit="KTS">        200.0 </vc>
  <gamma unit="DEG">       0.0 </gamma>
  <beta unit="DEG">        0.0 </beta>
  <running> -1 </running>
</initialize>
```

Or, if I'm scripting things with python I'll often use this option:

```python
# Set engines running
fdm['propulsion/engine[0]/set-running'] = 1
fdm['propulsion/engine[1]/set-running'] = 1
```

Or you could the same properties in a JSBSim xml script file as well.

**Reply by hudkinsnoah** [reply]
That did it! I was able to get the engines to actually start up and get running! One last question which is relating to the ground reactions. The x, y, and z locations that are given, how are those determined? Is there any resource in order to say exactly where to start the calculation from such as an origin point and then the common points that are used?

**Comment by seanmcleod** [comment]
Take a look at https://github.com/JSBSim-Team/jsbsim?tab=readme-ov-file#user-documentation and you'll see in the documentation a description of the various frames of reference. In particular for contact points they're specified in the structural frame.
